### PR TITLE
chore: upload metadata files required by Studio to the deployment bucket root

### PR DESCRIPTION
**Pull request by srquinn21**
#### Description of changes

**Problem:**
Amplify Studio is currently using the unzipper library to stream
metatdata files out of the zip file in the customer's deployment bucket
in order to retrieve data it needs for deployments and/or display in the
UI. However, this library is out of active development and has
significant error handling bugs that causing the Amplify Backend control
plane availability to degrade.

**Solution:**
Upload the needed metadata files to the s3 bucket outside of the zip
file so they are accessible with a single GetObject call and reaching
into the zip file is no longer required.

#### Issue #, if available
N/A

#### Description of how you validated changes

```
yarn setup-dev
amplify-dev init
amplify-dev add auth
amplify-dev push
```
Manually check the S3 console to see if files exist outside the zip

```
amplify-dev remove auth
amplify-dev push
```

Manually check the S3 console to see if files have been removed

#### Checklist

- [x] PR description included
- [x] `yarn test` passes

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-provider-awscloudformation/src/push-resources.ts [hunk 1]
```diff
@@ -555,6 +555,33 @@ export const updateStackForAPIMigration = async (context: $TSContext, category:
   }
 };
 
+/**
+ * Publish files that Amplify Studio depends on outside the zip file so that can read
+ * without sreaming from the zip.
```
Comment (jhockett) [inline packages/amplify-provider-awscloudformation/src/push-resources.ts:560]:
```suggestion
 * without streaming from the zip.
```

File: packages/amplify-provider-awscloudformation/src/push-resources.ts [hunk 2]
```diff
@@ -555,6 +555,33 @@ export const updateStackForAPIMigration = async (context: $TSContext, category:
...
+    'backend-config.json',
+    'schema.graphql',
+    'transform.conf.json',
+    'parameters.json',
+  ]
+    .flatMap(baseName => glob.sync(`**/${baseName}`, { cwd: amplifyDirPath }))
+    .filter(filePath => !filePath.startsWith('backend/'))
+    .map(filePath => ({
+      Body: fs.createReadStream(path.join(amplifyDirPath, filePath)),
+      Key: path.join(studioBackendDirName, filePath.replace('#current-cloud-backend/', '')),
```
Comment (jhockett) [inline packages/amplify-provider-awscloudformation/src/push-resources.ts:580]: Do all of these path operations work on Windows? I'm not sure if the forward slash will be handled properly

File: packages/amplify-provider-awscloudformation/src/push-resources.ts [hunk 3]
```diff
@@ -555,6 +555,33 @@ export const updateStackForAPIMigration = async (context: $TSContext, category:
...
+  const uploadFileParams = [
+    'cli.json',
+    'amplify-meta.json',
+    'backend-config.json',
+    'schema.graphql',
+    'transform.conf.json',
+    'parameters.json',
+  ]
+    .flatMap(baseName => glob.sync(`**/${baseName}`, { cwd: amplifyDirPath }))
+    .filter(filePath => !filePath.startsWith('backend/'))
```
Comment (edwardfoyle) [inline packages/amplify-provider-awscloudformation/src/push-resources.ts:577]: is the idea here to only grab files that are in `#current-cloud-backend`? And if so, why not only look in that directory to begin with?

File: packages/amplify-provider-awscloudformation/src/push-resources.ts [hunk 4]
```diff
@@ -555,6 +555,33 @@ export const updateStackForAPIMigration = async (context: $TSContext, category:
...
+  // Delete the contents of the studio backend directory first
+  await s3.deleteDirectory(bucketName, studioBackendDirName);
+  // Create a list of file params to upload to the deployment bucket
+  const uploadFileParams = [
+    'cli.json',
+    'amplify-meta.json',
+    'backend-config.json',
+    'schema.graphql',
+    'transform.conf.json',
+    'parameters.json',
```
Comment (edwardfoyle) [inline packages/amplify-provider-awscloudformation/src/push-resources.ts:574]: what values in parameters.json does studio use?

File: packages/amplify-provider-awscloudformation/src/push-resources.ts (hunk 2 above)
Comment (srquinn21) [reply packages/amplify-provider-awscloudformation/src/push-resources.ts:580]: Good catch. No need for the slashes as `path.join` normalizes. `glob` modules takes care of the os abstraction as well and requires forward slashes. Will rev for removing the others.

File: packages/amplify-provider-awscloudformation/src/push-resources.ts (hunk 3 above)
Comment (srquinn21) [reply packages/amplify-provider-awscloudformation/src/push-resources.ts:577]: Unfortunately we need the cli.json file as well which is outside `#current-cloud-backend`. The logic is easier this way than special casing that one file.

File: packages/amplify-provider-awscloudformation/src/push-resources.ts (hunk 4 above)
Comment (srquinn21) [reply packages/amplify-provider-awscloudformation/src/push-resources.ts:574]: We only use it for the storage and auth categories, but there are params in there about permissions and 3p idps that we need.

**Review by jhockett — APPROVED** [review]

**Review by edwardfoyle — APPROVED** [review]

**Pull request merged**